### PR TITLE
Fixing client not sending end

### PR DESCRIPTION
**Pull request by distantcam**
The client code was never sending the `END` message, while the server was in a state where it was reading from the stream. When the client disconnects the stream would throw a disconnect exception.

The tentacle acts as the "server" so this error would appear on the tentacle, which is then reported as an error in the running task.

Fixes OctopusDeploy/Issues#4266

### CODE REVIEW

File: source/Halibut/Transport/Protocol/MessageExchangeStream.cs [hunk 1]
```diff
@@ -321,8 +330,8 @@ void SetNormalTimeouts()
             if (!stream.CanTimeout)
                 return;
 
-            stream.WriteTimeout = (int) HalibutLimits.TcpClientSendTimeout.TotalMilliseconds;
-            stream.ReadTimeout = (int) HalibutLimits.TcpClientReceiveTimeout.TotalMilliseconds;
+            stream.WriteTimeout = (int)HalibutLimits.TcpClientSendTimeout.TotalMilliseconds;
+            stream.ReadTimeout = (int)HalibutLimits.TcpClientReceiveTimeout.TotalMilliseconds;
```
Comment (droyad) [inline source/Halibut/Transport/Protocol/MessageExchangeStream.cs]: Spaces again

File: source/Halibut/Transport/SecureClient.cs [hunk 2]
```diff
@@ -215,9 +211,9 @@ static TcpClient CreateTcpClient()
         {
             var client = new TcpClient(AddressFamily.InterNetworkV6)
             {
-                SendTimeout = (int) HalibutLimits.TcpClientSendTimeout.TotalMilliseconds,
-                ReceiveTimeout = (int) HalibutLimits.TcpClientReceiveTimeout.TotalMilliseconds,
-                Client = {DualMode = true}
+                SendTimeout = (int)HalibutLimits.TcpClientSendTimeout.TotalMilliseconds,
```
Comment (droyad) [inline source/Halibut/Transport/SecureClient.cs]: whitespace

File: source/Halibut/Transport/SecureListener.cs [hunk 3]
```diff
@@ -86,7 +86,7 @@ public int Start()
 
             log = logFactory.ForEndpoint(new Uri("listen://" + listener.LocalEndpoint));
             log.Write(EventType.ListenerStarted, "Listener started");
-            Task.Run(async () => await Accept());
+            Task.Run(Accept);
```
Comment (droyad) [inline source/Halibut/Transport/SecureListener.cs]: Why was this changed? This will not return until the first await is hit from my understanding

File: source/Halibut/Transport/SecureListener.cs [hunk 4]
```diff
@@ -113,7 +113,7 @@ async Task Accept()
                     {
                         log.WriteException(EventType.Error, "Error accepting TCP client", ex);
                         // Slow down the logs in case an exception is immediately encountered each AcceptTcpClientAsync
-                        Thread.Sleep(1000);
+                        await Task.Delay(1000);
```
Comment (droyad) [inline source/Halibut/Transport/SecureListener.cs]: Changing this would make it hard to see this loop in a debug dump

File: source/Halibut/Transport/ConnectionPool.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+using System.Runtime.Serialization;
+using Halibut.Diagnostics;
+using Halibut.Logging;
+
+namespace Halibut.Transport
+{
+    public class ConnectionPool<TKey, TPooledResource>
+        where TPooledResource : class, IPooledResource
+    {
+        readonly ConcurrentDictionary<TKey, ConcurrentDictionary<TPooledResource, byte>> pool = new ConcurrentDictionary<TKey, ConcurrentDictionary<TPooledResource, byte>>();
```
Comment (droyad) [inline source/Halibut/Transport/ConnectionPool.cs]: What's the purpose of this change? It seems unrelated.

File: source/Halibut.Tests/ProtocolFixture.cs [hunk 6]
```diff
@@ -407,8 +412,8 @@ public void Send<T>(T message)
 
             public T Receive<T>()
             {
-                output.AppendLine("<-- " + typeof(T).Name);     
-                return (T) (nextReadQueue.Count > 0 ? nextReadQueue.Dequeue() : default(T));
```
[the remark this answers is not in the document]
Comment (distantcam) [reply source/Halibut.Tests/ProtocolFixture.cs]: I suspect this is a resharper thing, as I don't have it so my formatter is removing the space.

File: source/Halibut/Transport/SecureListener.cs (hunk 3 above)
Comment (distantcam) [reply source/Halibut/Transport/SecureListener.cs]: This is creating an unnecessary async state machine and object. Task run starts a new thread to handle the call anyway, and then returns immediately.

Comment (distantcam) [reply source/Halibut/Transport/SecureListener.cs]: I think you're thinking of the yield state machine.

File: source/Halibut/Transport/ConnectionPool.cs (hunk 5 above)
Comment (distantcam) [reply source/Halibut/Transport/ConnectionPool.cs]: It's to fix this malarkey. https://github.com/OctopusDeploy/Halibut/blob/master/source/Halibut/Transport/SecureClientConnectionPool.cs#L51-L81

The initial problem is a concurrency issue where the same object could be returned twice, leading to it being added to the pool twice. This fixes the concurrency issue and ensures the item is only added once. That's why there's a new `ShouldNotAllowMultipleReturnsOfSameConnection` test.

This fix was needed because the pool was not cleaned up properly as part of the dispose, which meant one connection was not disconnecting when it should have.

**Comment by distantcam** [comment]
I'm not sure why this only occurs sometimes. I could only get this error to appear with a high concurrency of clients to the server, so I guess when there's a lot of tasks being sent to the tentacle in a concurrent mode?

File: source/Halibut/Transport/SecureListener.cs (hunk 4 above)
Comment (distantcam) [reply source/Halibut/Transport/SecureListener.cs]: I'm not super tied to this change. Potentially a thread sleep could cause a deadlock because it hogs a thread that could be returned to the async pool, but in practice I doubt that would happen anyway.

File: source/Halibut.Tests/ProtocolFixture.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (distantcam) [reply source/Halibut.Tests/ProtocolFixture.cs]: We should discuss using a standard formatting spec for our projects (editorconfig) but for now I don't want to hold this PR up. Should I change the spaces back or do you not care enough.

File: source/Halibut/Transport/ConnectionPool.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+        {
+            while (true)
+            {
+                var connections = pool.GetOrAdd(endPoint, i => new ConcurrentDictionary<TPooledResource, byte>());
+                AtomicTake(connections, out var connection);
+
+                if (connection == null || !connection.HasExpired())
+                    return connection;
+
+                connection.Dispose();
```
Comment (droyad) [inline source/Halibut/Transport/ConnectionPool.cs]: This should probably have a try/catch around it so that a faulted connection doesn't fail this new connection attempt

File: source/Halibut/Transport/ConnectionPool.cs [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+                if (AtomicTake(resources, out var dispose))
+                {
+                    dispose.Dispose();
+                }
+            }
+        }
+
+        public void Clear(TKey key, Diagnostics.ILog log = null)
+        {
+            if (!pool.TryRemove(key, out var connections))
```
Comment (droyad) [inline source/Halibut/Transport/ConnectionPool.cs]: What if another thread is on line 25 at this point. We would dispose all the connections and then the other thread would take an already disposed connection out of the pool. We could be halfway though the foreach on ln 56 when this occurs.

File: source/Halibut/Transport/SecureListener.cs (hunk 4 above)
Comment (droyad) [reply source/Halibut/Transport/SecureListener.cs]: Yeh I think we should keep it the way it is, it's super useful when debugging

File: source/Halibut/Transport/SecureListener.cs (hunk 3 above)
Comment (droyad) [reply source/Halibut/Transport/SecureListener.cs]: Ah yes, you are correct

Comment (droyad) [reply source/Halibut/Transport/SecureListener.cs]: I think we should change this back to `new Thread(Accept).Run()` now that .NET Core supports the Sync method.

File: source/Halibut/Transport/ConnectionPool.cs (hunk 7 above)
Comment (distantcam) [reply source/Halibut/Transport/ConnectionPool.cs]: `ConnectionPool` is generic and only works with `IPooledResource`. I'm not sure what you mean by faulted connection or how to detect that from the interface.

Comment (droyad) [reply source/Halibut/Transport/ConnectionPool.cs]: `IPooledResource` is really a `SecureConnection`. The Dispose method of that could throw an exception (although SocketException is already covered). All I meant was that a failure in this dispose call shouldn't cause the call to this method to fail.
